### PR TITLE
feat: Introduce more in-depth log collection for question rendering process

### PR DESCRIPTION
**Pull request by nwalters512**
This PR builds on #5410 by allowing us to construct a buffer of logs during the lifetime of a request and then emit them if the request takes more than 60 seconds to process. Ideally, we'd do this with OTel instrumentation, but the fact that we have to sample that means that we'd almost certainly not retain the interesting traces. This gives us a bunch of (hopefully) useful information:

```
error: ensureVariant transaction was not closed within 60 seconds {"authn_user_id":1,"course_instance_id":null,"group_work":false,"instance_question_id":null,"question_id":70,"user_id":1}
error: _selectQuestionWithClient complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:254:19","timestamp":"2022-02-23T22:03:28.116Z"}
error: getModule complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:116:19","timestamp":"2022-02-23T22:03:28.116Z"}
error: getContext complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1031:19","timestamp":"2022-02-23T22:03:28.124Z"}
error: getPythonCaller complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1043:21","timestamp":"2022-02-23T22:03:28.124Z"}
error: call complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:432:21","timestamp":"2022-02-23T22:03:28.506Z"}
error: execPythonServer complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:903:19","timestamp":"2022-02-23T22:03:28.506Z"}
error: processQuestionServer complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:976:23","timestamp":"2022-02-23T22:03:28.506Z"}
error: processQuestion complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1051:25","timestamp":"2022-02-23T22:03:28.507Z"}
error: returnPythonCaller complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1054:27","timestamp":"2022-02-23T22:03:28.507Z"}
error: generate complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:119:21","timestamp":"2022-02-23T22:03:28.507Z"}
error: getContext complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1074:19","timestamp":"2022-02-23T22:03:28.508Z"}
error: getPythonCaller complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1087:21","timestamp":"2022-02-23T22:03:28.508Z"}
error: processQuestionHtml complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:988:23","timestamp":"2022-02-23T22:03:29.256Z"}
error: call complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:432:21","timestamp":"2022-02-23T22:03:29.257Z"}
error: execPythonServer complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:903:19","timestamp":"2022-02-23T22:03:29.257Z"}
error: processQuestionServer complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1000:27","timestamp":"2022-02-23T22:03:29.258Z"}
error: processQuestion complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1095:25","timestamp":"2022-02-23T22:03:29.258Z"}
error: returnPythonCaller complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1098:27","timestamp":"2022-02-23T22:03:29.258Z"}
error: prepare complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:143:23","timestamp":"2022-02-23T22:03:29.258Z"}
error: _makeVariant complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:257:21","timestamp":"2022-02-23T22:03:29.258Z"}
error: _makeVariant complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:273:23","timestamp":"2022-02-23T22:03:29.266Z"}
error: _writeCourseIssues complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:288:27","timestamp":"2022-02-23T22:03:29.267Z"}
error: _makeAndInsertVariantWithClient complete {"data":{"variant_id":28},"location":"/Users/nathan/git/PrairieLearn/lib/question.js:374:23","timestamp":"2022-02-23T22:03:29.267Z"}
error: ensureVariant complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:1935:29","timestamp":"2022-02-23T22:03:29.268Z"}
error: select_submissions complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:1987:25","timestamp":"2022-02-23T22:03:29.276Z"}
error: getEffectiveQuestionType complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:2007:25","timestamp":"2022-02-23T22:03:29.276Z"}
error: processQuestionHtml complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:988:23","timestamp":"2022-02-23T22:03:30.023Z"}
error: call complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:432:21","timestamp":"2022-02-23T22:03:30.024Z"}
error: execPythonServer complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:903:19","timestamp":"2022-02-23T22:03:30.024Z"}
error: processQuestionServer complete {"location":"/Users/nathan/git/PrairieLearn/question-servers/freeform.js:1000:27","timestamp":"2022-02-23T22:03:30.024Z"}
error: render complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:2030:27","timestamp":"2022-02-23T22:03:30.026Z"}
error: select_issues complete {"location":"/Users/nathan/git/PrairieLearn/lib/question.js:2046:25","timestamp":"2022-02-23T22:03:30.029Z"}
```

With these logs in place, the hope is that we can better identify where exactly we get stuck while serving requests.

### CODE REVIEW

File: lib/code-caller.js [hunk 1]
```diff
@@ -147,7 +148,10 @@ class PythonCaller {
       paths: localOptions.paths,
     };
     const callDataString = JSON.stringify(callData);
-    this.callback = callback;
+    // This ensures that the callback is invoked in the same execution context
+    // that `call()` was invoked in. This allows `AsyncLocalStorage` to work
+    // correctly during asynchronous operations invoking `call()`.
+    this.callback = AsyncResource.bind(callback);
```
Comment (nwalters512) [inline lib/code-caller.js:154]: I thought this API was neat!

**Comment by nwalters512** [comment]
Ugh, the tests do not seem happy, I've had to cancel two runs now that took too long. I'll look into why they're failing tomorrow; hopefully I can reproduce locally.

File: packages/async-local-event-buffer/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  timestamp: Date;
+  message: string;
+  data?: any;
+}
+
+interface EventBuffer {
+  push: (msg: string, data?: any) => void;
+  flush: () => Message[];
+}
+
```
Comment (SethPoulsen) [inline packages/async-local-event-buffer/src/index.ts:13]: Seeing this stuff start to show up in the codebase makes me very happy.

**Comment by nwalters512** [comment]
We're pretty confident that we fixed the underlying issue in https://github.com/PrairieLearn/PrairieLearn/pull/5419, so I'm going to close out this PR for now. If this is ever needed again in the future, I know where to find it!

**Pull request closed without merging**